src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp

```diff
@@ -819,10 +784,15 @@ const char* ShenandoahGenerationalControlThread::gc_mode_name(ShenandoahGenerati
   }
 }
 
-void ShenandoahGenerationalControlThread::set_gc_mode(ShenandoahGenerationalControlThread::GCMode new_mode) {
+void ShenandoahGenerationalControlThread::set_gc_mode(GCMode new_mode) {
+  MonitorLocker ml(&_control_lock, Mutex::_no_safepoint_check_flag);
+  set_gc_mode(ml, new_mode);
+}
+
+void ShenandoahGenerationalControlThread::set_gc_mode(MonitorLocker& ml, GCMode new_mode) {
   if (_mode != new_mode) {
-    log_debug(gc)("Transition from: %s to: %s", gc_mode_name(_mode), gc_mode_name(new_mode));
-    MonitorLocker ml(&_regulator_lock, Mutex::_no_safepoint_check_flag);
+    log_debug(gc, thread)("Transition from: %s to: %s", gc_mode_name(_mode), gc_mode_name(new_mode));
+    EventMark event("Control thread transition from: %s, to %s", gc_mode_name(_mode), gc_mode_name(new_mode));
```

Comment:
As in my suggestions above for `notify_control_thread` and `notify_cancellation`, this might be an opportune time to do any sanity/consistency checks for `gc_mode` updates (wrt the other two `_request*` fields.